File: src/java/org/orbeon/saxon/dom4j/TypedNodeWrapper.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.saxon.dom4j;

import org.dom4j.Node;
import org.dom4j.QName;
import org.orbeon.oxf.xforms.InstanceData;
import org.orbeon.oxf.xforms.model.DataModel;
import org.orbeon.saxon.om.*;
import org.orbeon.saxon.trans.Err;
import org.orbeon.saxon.trans.XPathException;
import org.orbeon.saxon.type.AtomicType;
import org.orbeon.saxon.type.BuiltInType;
import org.orbeon.saxon.type.SchemaType;
import org.orbeon.saxon.type.Type;
import org.orbeon.saxon.value.StringValue;
import org.orbeon.saxon.value.UntypedAtomicValue;
import org.orbeon.saxon.value.Value;

/**
 * This wrapper is an extension of the Saxon node wrapper which is aware of XForms type annotations.
 */
public class TypedNodeWrapper extends NodeWrapper {

    protected TypedNodeWrapper(Object node, NodeWrapper parent, int index) {
        super(node, parent, index);
    }

    @Override
    protected NodeWrapper makeWrapper(Object node, DocumentWrapper docWrapper, NodeWrapper parent, int index) {
        return makeTypedWrapper(node, docWrapper, parent, index);
    }

    static NodeWrapper makeTypedWrapper(Object node, DocumentWrapper docWrapper, NodeWrapper parent, int index) {
        NodeWrapper wrapper;
        final Node dom4jNode = (Node) node;
        switch (dom4jNode.getNodeType()) {
            case Type.DOCUMENT:
                return docWrapper;
            case Type.ELEMENT:
            case Type.ATTRIBUTE:
            case Type.COMMENT:
            case Type.PROCESSING_INSTRUCTION:
            case Type.NAMESPACE:
            case Type.TEXT:
                wrapper = new TypedNodeWrapper(node, parent, index);
                wrapper.nodeKind = dom4jNode.getNodeType();
                break;
            case 4: // dom4j CDATA
                wrapper = new TypedNodeWrapper(node, parent, index);
                wrapper.nodeKind = Type.TEXT;
                break;
            default:
                throw new IllegalArgumentException("Bad node type in dom4j: " + node.getClass() + " instance " + node.toString());
        }

        wrapper.docWrapper = docWrapper;
        return wrapper;
    }

    @Override
    public SequenceIterator getTypedValue() throws XPathException {

        int annotation = getTypeAnnotation();
        if ((annotation & NodeInfo.IS_DTD_TYPE) != 0) {
            annotation = StandardNames.XS_UNTYPED_ATOMIC;
        }
        annotation &= NamePool.FP_MASK;
        if (annotation == -1 || annotation == StandardNames.XS_UNTYPED_ATOMIC || annotation == StandardNames.XS_UNTYPED) {
            return SingletonIterator.makeIterator(new UntypedAtomicValue(getStringValueCS()));
        } else {
            SchemaType stype = getConfiguration().getSchemaType(annotation);
            if (stype == null) {
                String typeName;
                try {
                    typeName = getNamePool().getDisplayName(annotation);
                } catch (Exception err) {
                    typeName = annotation + "";
                }
                throw new XPathException("Unknown type annotation " +
                        Err.wrap(typeName) + " in document instance");
            } else {
                return stype.getTypedValue(this);
            }
        }
    }

    @Override
    public Value atomize() throws XPathException {
        int annotation = getTypeAnnotation();
        if ((annotation & NodeInfo.IS_DTD_TYPE) != 0) {
            annotation = StandardNames.XS_UNTYPED_ATOMIC;
        }
        annotation &= NamePool.FP_MASK;
        if (annotation == -1 || annotation == StandardNames.XS_UNTYPED_ATOMIC || annotation == StandardNames.XS_UNTYPED) {
            return new UntypedAtomicValue(getStringValueCS());
        } else {
            SchemaType stype = getConfiguration().getSchemaType(annotation);
            if (stype == null) {
                String typeName;
                try {
                    typeName = getNamePool().getDisplayName(annotation);
                } catch (Exception err) {
                    typeName = annotation + "";
                }
                throw new XPathException("Unknown type annotation " +
                        Err.wrap(typeName) + " in document instance");
            } else {
                return stype.atomize(this);
            }
        }
    }

    @Override
    public int getTypeAnnotation() {

        final QName nodeType = InstanceData.getType((Node) node);
        if (nodeType == null) {
            return getUntypedType();
        } else {
            // Extract QName
            final String uri = nodeType.getNamespaceURI();
            final String localname = nodeType.getName();

            final int requestedTypeFingerprint = StandardNames.getFingerprint(uri, localname);
            if (requestedTypeFingerprint == -1) {
                // Back to default case
                return getUntypedType();
            } else {
                // Return identified type
                // NOTE: Return a type iif the value matches the type, because that's required by the XPath semantic.
                final StringValue value = new StringValue(DataModel.getValue(this));

                try {
                    value.convert((AtomicType) BuiltInType.getSchemaType(requestedTypeFingerprint), getConfiguration().getConversionContext());
                    // Value matches type, return it
                    return requestedTypeFingerprint;
                } catch (XPathException e) {
                    // Back to default case
                    return getUntypedType();
                }
            }
        }
    }

    private int getUntypedType() {
        if (getNodeKind() == Type.ATTRIBUTE) {
            return StandardNames.XS_UNTYPED_ATOMIC;
        }
        return StandardNames.XS_UNTYPED;
    }
}

File: xforms-runtime/js/src/main/scala/org/orbeon/dom/saxon/TypedNodeWrapper.scala
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.dom.saxon

import org.orbeon.dom
import org.orbeon.dom.Node
import org.orbeon.oxf.xforms.analysis.model.ModelDefs
import org.orbeon.oxf.xforms.model.InstanceData
import org.orbeon.oxf.xml.XMLConstants
import org.orbeon.saxon.model.{BuiltInAtomicType, SchemaType, Type, Untyped}
import org.orbeon.saxon.om._
import org.orbeon.saxon.value.UntypedAtomicValue
import org.orbeon.xforms.XFormsNames

import scala.util.control.NonFatal


/**
 * This wrapper is an extension of the Saxon node wrapper which is aware of XForms type annotations.
 */
object TypedNodeWrapper {

  private[saxon] def makeTypedWrapper(node: Node, docWrapper: DocumentWrapper, parent: NodeWrapper): NodeWrapper =
    if (node.isInstanceOf[dom.Document])
      docWrapper
    else
      new TypedNodeWrapper(node, docWrapper, parent)

  class TypedValueException(val nodeName: String, val typeName: String, val nodeValue: String) extends RuntimeException
}

class TypedNodeWrapper private (node: dom.Node, docWrapper: DocumentWrapper, parent: NodeWrapper)
  extends ConcreteNodeWrapper(node, docWrapper, parent) {

  override protected def makeWrapper(node: dom.Node, docWrapper: DocumentWrapper, parent: NodeWrapper): NodeWrapper =
    TypedNodeWrapper.makeTypedWrapper(node, docWrapper, parent)

  override def atomize(): AtomicSequence =
    checkTypeAnnotation(new UntypedAtomicValue(_), _.atomize(this))

  override def getSchemaType: SchemaType = {
    val nodeType = InstanceData.getType(node)
    if (nodeType == null) {
      getUntypedType
    } else {
      // Extract QName
      var uri       = nodeType.namespace.uri
      val localname = nodeType.localName

      // For type annotation purposes, `xf:integer` is translated into `xs:integer`. This is because XPath has no
      // knowledge of the XForms union types.
      if (uri == XFormsNames.XFORMS_NAMESPACE_URI && ModelDefs.XFormsVariationTypeNames(localname))
        uri = XMLConstants.XSD_URI

      val schemaType = getConfiguration.getSchemaType(new StructuredQName(nodeType.namespace.prefix, uri, localname))

      if (schemaType eq null)
        getUntypedType // back to default case
      else
        schemaType // return identified type
    }
  }

  private def checkTypeAnnotation[A](untyped: CharSequence => A, typed: SchemaType => A): A = {

    val schemaType = getSchemaType

    if (schemaType == null || schemaType == BuiltInAtomicType.UNTYPED_ATOMIC || schemaType == Untyped.getInstance) {
      untyped(getStringValueCS)
    } else {
//      val stype = getConfiguration.getSchemaType(schemaType)
//      if (stype eq null)
//        throw new XPathException(s"Unknown type annotation `${Err.wrap(getAnnotationTypeName(schemaType))}` in document")
//      else
      try
        typed(schemaType)
      catch {
        case NonFatal(_) =>
          // TODO: Would be good to pass err.getMessage()
          throw new TypedNodeWrapper.TypedValueException(getDisplayName, schemaType.getDisplayName, getStringValue)
      }
    }
  }

  private def getUntypedType: SchemaType =
    getNodeKind match {
      case Type.ATTRIBUTE               => BuiltInAtomicType.UNTYPED_ATOMIC
      case Type.DOCUMENT | Type.ELEMENT => Untyped.getInstance
      case _                            => null
    }
}